lib/collection/src/collection.rs

```diff
@@ -574,20 +574,40 @@ impl Collection {
             .filter(|result| matches!(result, Err(_)))
             .count();
 
+        // one request per shard
+        let result_len = results.len();
+
         if with_error > 0 {
-            let err = results
+            let first_err = results
                 .into_iter()
                 .find(|result| matches!(result, Err(_)))
                 .unwrap();
-            let num_shards = self.shards_holder.read().await.len();
-            if with_error < num_shards {
-                err.map_err(|err| CollectionError::InconsistentFailure {
-                    shards_total: num_shards as u32,
-                    shards_failed: with_error as u32,
-                    first_err: format!("{err}"),
+            // inconsistent if only a subset of the requests fail - one request per shard.
+            if with_error < result_len {
+                first_err.map_err(|err| {
+                    // compute final status code based on the first error
+                    // e.g. a partially successful batch update failing because of bad input is a client error
+                    let status_code = match err {
+                        CollectionError::BadInput { .. } => 400,
+                        CollectionError::BadRequest { .. } => 400,
+                        CollectionError::BadShardSelection { .. } => 500,
+                        CollectionError::NotFound { .. } => 404,
+                        CollectionError::PointNotFound { .. } => 404,
+                        CollectionError::ServiceError { .. } => 500,
+                        CollectionError::Cancelled { .. } => 500,
+                        CollectionError::InconsistentFailure { .. } => 500,
```

Comment:
Don't we already have these conversion somewhere? I mean just not to duplicate the code and have accidentally different status codes

Reply:
> Don't we already have these conversion somewhere? I mean just not to duplicate the code and have accidentally different status codes

First we have first a conversion from `CollectionError` to `StorageError` via a `From` impl
And then we have the mapping between `StorageError` and Http response.

I have found a [helper](https://github.com/qdrant/qdrant/blob/06afc5234e2629ab7a48870010a0fae08d40a74a/src/actix/helpers.rs#L10) for this but it can't be used directly from `collection` crate.

Reply:
Maybe then it is better to just store `CollectionError` instead of status. Then on conversion from `CollectionError` to `StorageError` also convert the stored error and then use it in the API for the propper status. What do you think?